pkg/scheduler/framework/plugins/nodeports/node_ports.go

```diff
@@ -113,7 +113,8 @@ func (pl *NodePorts) EventsToRegister() []framework.ClusterEvent {
 func (pl *NodePorts) Filter(ctx context.Context, cycleState *framework.CycleState, pod *v1.Pod, nodeInfo *framework.NodeInfo) *framework.Status {
 	wantPorts, err := getPreFilterState(cycleState)
 	if err != nil {
-		return framework.AsStatus(err)
+		// Fallback to calculate containerPorts here when PreFilter is disabled.
+		wantPorts = getContainerPorts(pod)
```

Comment:
It's not reasonable. Why you want to disable the PreFilter？

Reply:
I'm not trying to disable prefilter here, just a fallback calculation when getPreFilterState fails.
Similar processing like nodeaffinity filter:
https://github.com/kubernetes/kubernetes/blob/5108b0a3a0a9622a69fcd1fb3d78de77e6c3d7cb/pkg/scheduler/framework/plugins/nodeaffinity/node_affinity.go#L154-L159

Reply:
I am not opposed to this change, but... should we implement fallback like this in all plugins?
Some other plugins currently don't do the fallback.
Is there any case that users want to disable PreFilter but enable Filter?

Reply:
I'd say a plugin's PreFilter and Filter should be enabled/disabled altogether, although technically it's not enforced. Can we think of a valid use case to only enable PreFilter/Filter of a plugin? If not, we probably can implement enforced semantics.

Reply:
Yeah, I agree with force using PreFilter and Filter together.
Given some plugins currently don't do the fallback like this PR and we haven't got any issue for that so far (AFAIK), I think probably few people are using Filter plugin without PreFIlter phase. 
And enforcing it could be useful in terms of preventing users from mistakes in the configuration of Plugins.

Reply:
Created https://github.com/kubernetes/kubernetes/issues/111309 to discuss.

Reply:
The reason we have optional PreFilter in some plugins is that we added the PreFilter phase much later. Then, you could have clusters upgrading k8s versions that have a custom configuration that would start failing.

I don't think We should add this fallback retroactively to plugins that had PreFilter from the beginning.

Reply:
So this is a historic problem left, I'd like to push this forward by:

- [ ] Enable PreFilter for every in-tree plugins
- [ ] Tie the `PreFilter` and `Filter` together
- [ ] Remove the fallback path if they have

Make sense?

Reply:
1. PreFilter is enabled by default for every in-tree plugin
2. The current recommendation is to use the multipoint API, which enables all the extension points of a plugin, but
3. You can still use the individual extension points of PreFilter and Filter

For backwards compatibility, we should keep things like they are.... Maybe we could start writing Warnings when using a Filter without it's corresponding PreFilter and eventually remove the fallback path.

Reply:
Right, we should leave things as is. 

The whole idea behind multipoint is to simplify plugin configuration: users shouldn't really think about extension points in general, they should think of plugins, and multipoint allows that mode of operation.

We left the older configuration approach for backward compatibility and to allow advanced users more flexibility.

I think we removed all fallback logic long time ago.

Reply:
To clarify more clearly:
`Tie the PreFilter and Filter together` is not mean combine two extension points into one, I mean they should be enabled/disabled together. Bad expressions.

> Maybe we could start writing Warnings when using a Filter without it's corresponding PreFilter and eventually remove the fallback path

+1 add warning to the plugins which has fallback path as a start(plugins with no fallback path now will return error directly). And finally let's remove the fallback path and the warning.

> The whole idea behind multipoint is to simplify plugin configuration: users shouldn't really think about extension points in general, they should think of plugins, and multipoint allows that mode of operation.

Absolutely.

> I think we removed all fallback logic long time ago.

How about this one? https://github.com/kubernetes/kubernetes/blob/5f6b60bb1323a60f59119e315ac52b8a453a3939/pkg/scheduler/framework/plugins/nodeaffinity/node_affinity.go#L154-L159

Reply:
In any case, this PR shouldn't merge as is.

Reply:
> How about this one?
> [kubernetes/pkg/scheduler/framework/plugins/nodeaffinity/node_affinity.go](https://github.com/kubernetes/kubernetes/blob/5f6b60bb1323a60f59119e315ac52b8a453a3939/pkg/scheduler/framework/plugins/nodeaffinity/node_affinity.go#L154-L159)

I am guessing this was done for backward compatibility when PreFilter was introduced to NodeAffinity (for example configurations that reordered NodeAffinity). Things like this (adding a PreFilter to an existing Filter) was what motivated the multi-point configuration approach. In fact I initially opposed continuing to have the finer grained approach (explicitly listing extensions in the CC), but the community wanted to continue to have it, IIRC mostly to allow easier development/experimentation.

Reply:
Some updates, in other projects, like volcano, it will use the `Filter` extension point for predicates but without `PreFilter`, so I'm considering it this is our responsibility to support this in kubernetes upstream.